## Business Problems:

## Business Problem Statement for Machine Learning
**How to segment the customers at Olist marketplace so we can divide customers based on their shopping behaviour?**
- Value	: Customer Segmentation
- Goals	: Divide customers based on their shopping behaviour

## Business Problem Statement for Analytics
**What kind of treatment for each cluster to increase retention rate customer?**
- Value	: Give treatment for each cluster
- Goals	: Increase retention rate customer


In [ ]:
# Data
import pandas as pd
import datetime as dt

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Data Understanding:
- Data mengenai perusahaan *E-commerce* di Brazil, yaitu perusahaan *Olist*
- Data ini diambil dari tahun 2016 - 2018 dan disebarkan secara public di <a href = "https://www.kaggle.com/olistbr/brazilian-ecommerce">Kaggle</a> pada tahun 2018 bulan September
- Data-data yang ada berisi tentang transaksi *customer* dan *seller*
- Karena yang ingin dilakukan adalah *customer segmentation* maka hanya akan diambil data yang mengenai *customer* saja
- Ada 116584 data transaksi

### Merge Data:
Dikarenakan data yang berhubungan dengan transaksi *customer* terpisah-pisah, hal pertama yang harus dilakukan adalah mengabungkan data-data tersebut.

<p> <p>
<figure>
  <center><img src = "https://i.imgur.com/HRhd2Y0.png" alt ="Scheme Data Table" width = 700; height = 500></center>
  <figcaption align = "center">Gambar 1. <em>Scheme Data</em> </figcaption>
</figure>
<p> <p>

In [ ]:
# Read data yang dibutuhkan
olist_order = pd.read_csv('/content/drive/MyDrive/Project Purwadhika/Dataset/olist_orders_dataset.csv')
olist_item = pd.read_csv('/content/drive/MyDrive/Project Purwadhika/Dataset/olist_order_items_dataset.csv')
olist_payment = pd.read_csv('/content/drive/MyDrive/Project Purwadhika/Dataset/olist_order_payments_dataset.csv')
olist_product = pd.read_csv('/content/drive/MyDrive/Project Purwadhika/Dataset/olist_products_dataset.csv')
olist_customer = pd.read_csv('/content/drive/MyDrive/Project Purwadhika/Dataset/olist_customers_dataset.csv')
olist_review = pd.read_csv('/content/drive/MyDrive/Project Purwadhika/Dataset/olist_order_reviews_dataset.csv')
olist_translate = pd.read_csv('/content/drive/MyDrive/Project Purwadhika/Dataset/product_category_name_translation.csv')

In [ ]:
# Merge data menjadi satu 
df_olist = pd.merge(olist_order, olist_item, on='order_id', how='left')
df_olist = pd.merge(df_olist, olist_product, on='product_id', how='inner')
df_olist = pd.merge(df_olist, olist_payment, on='order_id', how = 'left')
df_olist = pd.merge(df_olist, olist_review, on='order_id', how='left')
df_olist = pd.merge(df_olist, olist_customer, on='customer_id', how='right')
df_olist = pd.merge(df_olist, olist_translate, on='product_category_name', how='inner')

df_olist.head()

order_id  ... product_category_name_english
0  00e7ee1b050b8499577073aeb2a297a1  ...              office_furniture
1  b2059ed67ce144a36e2aa97d2c9e9ad2  ...              office_furniture
2  951670f92359f4fe4a63112aa7306eba  ...              office_furniture
3  997fd4b37386d10b57d4ac8cdec011a6  ...              office_furniture
4  b63fff1ef8b6201df68bb611ee41129e  ...              office_furniture

[5 rows x 37 columns]

In [ ]:
print("Jumlah row yang dimiliki: ", len(df_olist))

Jumlah row yang dimiliki:  116584


In [ ]:
df_olist.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date',
       'price', 'freight_value', 'product_category_name',
       'product_name_lenght', 'product_description_lenght',
       'product_photos_qty', 'product_weight_g', 'product_length_cm',
       'product_height_cm', 'product_width_cm', 'payment_sequential',
       'payment_type', 'payment_installments', 'payment_value', 'review_id',
       'review_score', 'review_comment_title', 'review_comment_message',
       'review_creation_date', 'review_answer_timestamp', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'product_category_name_english'],
      dtype='object')

### Drop columns:
Setelah penggabungan data dari beberapa tables, kami akan menghapus beberapa column yang tidak ada kaitannya dengan transaksi *behavior* dari *customer* 

In [ ]:
df_olist_clean = df_olist.drop(columns=['order_status', 'order_approved_at', 'order_delivered_customer_date', 'order_delivered_carrier_date', 'order_delivered_customer_date', 
                                  'order_estimated_delivery_date', 'seller_id', 'shipping_limit_date','product_category_name', 'product_name_lenght', 
                                  'product_description_lenght', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm',
                                  'payment_installments','review_id', 'review_comment_title', 'review_comment_message', 'review_creation_date', 
                                  'review_answer_timestamp', 'customer_id'])

In [ ]:
df_olist_clean.head()

order_id  ... product_category_name_english
0  00e7ee1b050b8499577073aeb2a297a1  ...              office_furniture
1  b2059ed67ce144a36e2aa97d2c9e9ad2  ...              office_furniture
2  951670f92359f4fe4a63112aa7306eba  ...              office_furniture
3  997fd4b37386d10b57d4ac8cdec011a6  ...              office_furniture
4  b63fff1ef8b6201df68bb611ee41129e  ...              office_furniture

[5 rows x 16 columns]

## Attribute Information:
<table>
  <tr>
    <th>Attribute</th>
    <th>Data Type</th>
    <th>Description</th>
    <th>Unit Analysis</th>
  </tr>
  <tr>
    <td>order_id</td>
    <td>object</td>
    <td>Unik ID dari order</td>
    <td>Setiap baris merepresentasikan ID order yang yang muncul ketika customer
    melakukan order</td>
  </tr>
  <tr>
    <td>order_purchase_stamp</td>
    <td>datetime64[s]</td>
    <td>Waktu pembelian customer</td>
    <td>Setiap baris merepresentasikan waktu ketika customer melakukan order</td>
  </tr>
  <tr>
    <td>order_item_id</td>
    <td>float64</td>
    <td>Jumlah barang per-order</td>
    <td>Setiap baris merepresentasikan jumlah barang yang dibeli customer dalam sekali order</td>
  </tr>
  <tr>
    <td>product_id</td>
    <td>object</td>
    <td>ID dari sebuah product</td>
    <td>Setiap baris merepresentasikan ID dari sebuah product yang diorder oleh customer</td>
  </tr>
  <tr>
    <td>price</td>
    <td>float64</td>
    <td>Harga dari sebuah product</td>
    <td>Setiap baris merepresentasikan harga dari sebuah product yang diorder oleh customer</td>
  </tr>
  <tr>
    <td>freight_value</td>
    <td>float64</td>
    <td>Nilai pengiriman product</td>
    <td>Setiap baris merepresentasikan nilai pengirim dari sebuah product yang diorder oleh customer. Apabila item lebih dari 1, nilainya dipisah</td>
  </tr>
  <tr>
    <td>product_photo_qty</td>
    <td>float64</td>
    <td>Jumlah foto sebuah product</td>
    <td>Setiap baris merepresentasikan jumlah foto dari sebuah product yang diiklankan</td>
  </tr>
  <tr>
    <td>payment_sequential</td>
    <td>float64</td>
    <td>Jumlah payment method</td>
    <td>Setiap baris merepresentasikan jumlah metode payment yang dilakukan oleh customer. Customer dapat membayar sebuah transaksi lebih dari 1 metode</td>
  </tr>
  <tr>
    <td>payment_type</td>
    <td>object</td>
    <td>Metode pembayaran</td>
    <td>Setiap baris merepresentasikan metode yang digunakan customer dalam membayar sebuah transaksi</td>
  </tr>
  <tr>
    <td>payment_value</td>
    <td>float64</td>
    <td>Nilai Transaksi</td>
    <td>Setiap baris merepresentasikan nilai transaksi yang harus dibayar customer</td>
  </tr>
  <tr>
    <td>review_score</td>
    <td>float64</td>
    <td>Tingkat kepuasan customer</td>
    <td>Setiap baris merepresentasikan tingkat kepuasan customer dalam bertransaksi, nilainya dari 1 - 5</td>
  </tr>
  <tr>
    <td>customer_unique_id</td>
    <td>object</td>
    <td>Unik ID dari customer</td>
    <td>Setiap baris merepresentasikan ID customer yang melakukan
    order</td>
  </tr>
  <tr>
    <td>customer_zip_code_prefix</td>
    <td>int64</td>
    <td>Kode pos dari customer</td>
    <td>Setiap baris merepresentasikan kode pos dari tempat tinggal customer, namun hanya 5 kode awal</td>
  </tr>
  <tr>
    <td>customer_city</td>
    <td>object</td>
    <td>Nama kota dari customer</td>
    <td>Setiap baris merepresentasikan nama kota tempat tinggal customer</td>
  </tr>
  <tr>
    <td>customer_state</td>
    <td>object</td>
    <td>Negara Bagian customer</td>
    <td>Setiap baris merepresentasikan negara bagian dari tempat tinggal customer</td>
  </tr>
  <tr>
    <td>product_category_name_english</td>
    <td>object</td>
    <td>Kategori product</td>
    <td>Setiap baris merepresentasikan nama kategori dari sebuah product dalam Bahasa Inggris</td>
  </tr>
  <tr>
    <td>month_order</td>
    <td>object</td>
    <td>Nama bulan order dilakukan</td>
    <td>Setiap baris merepresentasikan nama bulan dari tanggal pembelian dilakukan</td>
  </tr>
  <tr>
    <td>weekday_order</td>
    <td>object</td>
    <td>Nama hari order dilakukan</td>
    <td>Setiap baris merepresentasikan nama hari dari tanggal pembelian dilakukan</td>
  </tr>
  <tr>
    <td>month_year_order</td>
    <td>period[M]</td>
    <td>Tahun dan bulan pembelian</td>
    <td>Setiap baris merepresentasikan tahun dan bulan dari tanggal pembelian dilakukan</td>
  </tr>

</table>

# Preprocessing Part 1:

### Handling Missing Value:

In [ ]:
df_olist_clean.isnull().sum()

order_id                         0
order_purchase_timestamp         0
order_item_id                    0
product_id                       0
price                            0
freight_value                    0
product_photos_qty               0
payment_sequential               3
payment_type                     3
payment_value                    3
review_score                     0
customer_unique_id               0
customer_zip_code_prefix         0
customer_city                    0
customer_state                   0
product_category_name_english    0
dtype: int64

In [ ]:
df_olist_clean[df_olist_clean['payment_type'].isnull()]

order_id  ... product_category_name_english
51505  bfbd0f9bdef84302105ad712db648a6c  ...                 health_beauty
51506  bfbd0f9bdef84302105ad712db648a6c  ...                 health_beauty
51507  bfbd0f9bdef84302105ad712db648a6c  ...                 health_beauty

[3 rows x 16 columns]

Column "payment_sequential", "payment_type", dan "payment_value" terdapat 3 data yang *missing value*. *Missing value* pada column tersebut tergolong dalam *Missing Value Completely at Random (MCAR)* yang artinya *missing data* yang terjadi benar-benar secara *random* dan tidak berkaitan dengan variable lain maupun dirinya sendiri. Kemungkinan hal ini terjadi karena adanya kesalahan dalam meng-*input* data. Untuk mengatasi hal tersebut, kami memutuskan untuk men-*drop* ketiga data tersebut dikarenakan jumlah data yang kami punya juga cukup besar. 

In [ ]:
df_olist_clean.dropna(inplace = True)

In [ ]:
df_olist_clean.isnull().sum()

order_id                         0
order_purchase_timestamp         0
order_item_id                    0
product_id                       0
price                            0
freight_value                    0
product_photos_qty               0
payment_sequential               0
payment_type                     0
payment_value                    0
review_score                     0
customer_unique_id               0
customer_zip_code_prefix         0
customer_city                    0
customer_state                   0
product_category_name_english    0
dtype: int64

### Handling Duplicate Value:
Proses selanjutnya adalah kita melihat apakah data yang sudah kita *clean* ada data yang *duplicate* atau tidak.

In [ ]:
df_olist_clean.duplicated().sum()

414

In [ ]:
df_olist_clean[df_olist_clean.duplicated()].head()

order_id  ... product_category_name_english
53    5040757d4e06a4be96d3827b860b4e7c  ...              office_furniture
1254  cb5b7fb63c0add7098c025cb737df02e  ...              office_furniture
2706  712cb14ddeca38aaeb49dea61f811018  ...                    housewares
2930  9de38c42ea4d33af19aac6c4d6c4b146  ...                    housewares
3383  cba44aa41766f6210b875f576a4241e4  ...                    housewares

[5 rows x 16 columns]

Dari hasil diatas, ditemukan ada 414 data yang memiliki nilai yang benar-benar sama. Pada umumnya, *duplicate value* perlu dihapus, namun sebelum dihapus kami akan melakukan pengecekan pada data yang belum diproses untuk mengetahui apakah data yang *duplicate* ini memiliki arti atau tidak.

In [ ]:
df_olist[(df_olist['order_id'] == "5040757d4e06a4be96d3827b860b4e7c") | (df_olist['order_id'] == "cb5b7fb63c0add7098c025cb737df02e")]

order_id  ... product_category_name_english
52    5040757d4e06a4be96d3827b860b4e7c  ...              office_furniture
53    5040757d4e06a4be96d3827b860b4e7c  ...              office_furniture
1253  cb5b7fb63c0add7098c025cb737df02e  ...              office_furniture
1254  cb5b7fb63c0add7098c025cb737df02e  ...              office_furniture

[4 rows x 37 columns]

Hasil pengecekan menunjukan bahwa data yang *duplicate* terjadi karena adanya *customer* yang memberikan 2 *review* terhadap transaksi yang sama. Berdasarkan pengecekan tersebut, kami memutuskan untuk menghapus data yang *duplicate* karena transaksi yang terjadi sebenarnya sama.

In [ ]:
df_olist_clean = df_olist_clean.drop_duplicates()

### Casting Data Types dan Feature Engineering:
*Formating Data Types* dilakukan untuk memastikan bahwa *data types* yang kami punya sesuai dengan yang seharusnya. Apabila tidak sesuai, maka *data types* dari sebuah *columns* harus diubah. Selain itu, *Formatting Data Types* juga digunakan untuk memastikan bahwa value dari *data categorical* sudah konsisten, tidak ada nama yang memiliki arti yang sama, namun *categorical*-nya berbeda

In [ ]:
# Cek tipe data tiap column
df_olist_clean.dtypes

order_id                          object
order_purchase_timestamp          object
order_item_id                    float64
product_id                        object
price                            float64
freight_value                    float64
product_photos_qty               float64
payment_sequential               float64
payment_type                      object
payment_value                    float64
review_score                     float64
customer_unique_id                object
customer_zip_code_prefix           int64
customer_city                     object
customer_state                    object
product_category_name_english     object
dtype: object

Dari hasil pengecekan, *columns* "order_purchase_timestamp" akan lebih baik apabila *data types*-nya diubah menjadi *datetime*. Selain mengubah *data types*, kami juga menambahkan beberapa *column* baru dari hasil ekstraksi *column* "order_purchase_timestamp" yang kemungkinan dapat digunakan pada proses *EDA*

In [ ]:
# Convert type data tanggal jadi datetime
df_olist_clean['order_purchase_timestamp'] = pd.to_datetime(df_olist_clean['order_purchase_timestamp'])
df_olist_clean['order_purchase_timestamp'].dt.strftime('%Y-%m-%d')

df_olist_clean['month_order'] = df_olist_clean['order_purchase_timestamp'].dt.month_name()
df_olist_clean['weekday_order'] = df_olist_clean['order_purchase_timestamp'].dt.day_name()
df_olist_clean['month_year_order'] = df_olist_clean['order_purchase_timestamp'].dt.to_period('M')
df_olist_clean['date_order'] = df_olist_clean['order_purchase_timestamp'].dt.day

df_olist_clean.head()

order_id  ... date_order
0  00e7ee1b050b8499577073aeb2a297a1  ...         16
1  b2059ed67ce144a36e2aa97d2c9e9ad2  ...         19
2  951670f92359f4fe4a63112aa7306eba  ...         13
3  997fd4b37386d10b57d4ac8cdec011a6  ...          8
4  b63fff1ef8b6201df68bb611ee41129e  ...          2

[5 rows x 20 columns]

In [ ]:
df_olist_clean.dtypes

order_id                                 object
order_purchase_timestamp         datetime64[ns]
order_item_id                           float64
product_id                               object
price                                   float64
freight_value                           float64
product_photos_qty                      float64
payment_sequential                      float64
payment_type                             object
payment_value                           float64
review_score                            float64
customer_unique_id                       object
customer_zip_code_prefix                  int64
customer_city                            object
customer_state                           object
product_category_name_english            object
month_order                              object
weekday_order                            object
month_year_order                      period[M]
date_order                                int64
dtype: object

### Inconsisten Variable:
Untuk memastikan bahwa data categorical nilainya konsisten, kami akan mengecek data-data categorical yang kami punya. Data Categorical yang kami punya adalah "payment_type", "customer_city", "customer_state", "product_category_name_english"

In [ ]:
# Cek categorical data
categorical_data = ["payment_type", "customer_city", "customer_state", "product_category_name_english"]

for i in categorical_data:
  print(f"Unique Value dari {i}")
  print(sorted(df_olist_clean[i].unique()))
  print('\n')

Unique Value dari payment_type
['boleto', 'credit_card', 'debit_card', 'voucher']


Unique Value dari customer_city
['abadia dos dourados', 'abadiania', 'abaete', 'abaetetuba', 'abaiara', 'abaira', 'abare', 'abatia', 'abdon batista', 'abelardo luz', 'abrantes', 'abre campo', 'abreu e lima', 'acaiaca', 'acailandia', 'acajutiba', 'acarau', 'acari', 'acegua', 'acopiara', 'acreuna', 'acu', 'acucena', 'adamantina', 'adhemar de barros', 'adolfo', 'adrianopolis', 'adustina', 'afogados da ingazeira', 'afonso claudio', 'afranio', 'agisse', 'agrolandia', 'agronomica', 'agua boa', 'agua branca', 'agua clara', 'agua comprida', 'agua doce', 'agua doce do norte', 'agua fria de goias', 'agua limpa', 'agua nova', 'agua preta', 'agua santa', 'aguai', 'aguas belas', 'aguas claras', 'aguas da prata', 'aguas de lindoia', 'aguas de santa barbara', 'aguas de sao pedro', 'aguas formosas', 'aguas frias', 'aguas lindas de goias', 'aguas mornas', 'aguas vermelhas', 'agudo', 'agudos', 'aguia branca', 'aimores', 

Bila diperhatikan, pada column "product_category_name_english" ada nilai yang memiliki 2 versi penamaan, seperti "home_appliances" dan "home_appliances_2", "home_confort" dan "home_comfort_2" sehingga kedua kategori tersebut akan dibuat sama.

In [ ]:
def format_val(x):
  if x == 'home_appliances_2':
    return "home_appliances"
  elif x == 'home_confort':
    return "home_comfort"
  elif x == "home_comfort_2":
    return "home_comfort"
  
  return x

df_olist_clean['product_category_name_english'] = df_olist_clean['product_category_name_english'].apply(format_val)

In [ ]:
print("Unique value dari product_category_name_english:\n")
print(sorted(df_olist_clean['product_category_name_english'].unique()))

Unique value dari product_category_name_english:

['agro_industry_and_commerce', 'air_conditioning', 'art', 'arts_and_craftmanship', 'audio', 'auto', 'baby', 'bed_bath_table', 'books_general_interest', 'books_imported', 'books_technical', 'cds_dvds_musicals', 'christmas_supplies', 'cine_photo', 'computers', 'computers_accessories', 'consoles_games', 'construction_tools_construction', 'construction_tools_lights', 'construction_tools_safety', 'cool_stuff', 'costruction_tools_garden', 'costruction_tools_tools', 'diapers_and_hygiene', 'drinks', 'dvds_blu_ray', 'electronics', 'fashio_female_clothing', 'fashion_bags_accessories', 'fashion_childrens_clothes', 'fashion_male_clothing', 'fashion_shoes', 'fashion_sport', 'fashion_underwear_beach', 'fixed_telephony', 'flowers', 'food', 'food_drink', 'furniture_bedroom', 'furniture_decor', 'furniture_living_room', 'furniture_mattress_and_upholstery', 'garden_tools', 'health_beauty', 'home_appliances', 'home_comfort', 'home_construction', 'houseware

# Save Data:
Setelah proses membersihkan data yang missing dan duplicate, kami akan menyimpan data tersebut agar dapat digunakan kembali. Pembersihan outlier akan dilakukan pada step EDA karena untuk mendeteksi outlier kami perlu melakukan explorasi datanya terlebih dahulu agar kami tidak sembarangan menghapus outlier

In [ ]:
df_olist_clean.to_csv('/content/drive/MyDrive/Project Purwadhika/data_outlist_clean.csv', index = False, date_format = '%Y-%m-%d')